### PR TITLE
[OCPCLOUD-1354] Add Network Interface Type to AWS Machine Provider Spec

### PR DESCRIPTION
**Pull request by JoelSpeed**
This PR adds the API to allow users to choose between standard network interfaces and AWS Elastic Fabric Adapter interfaces for the EC2 instance created by the Machine.

This is related to https://issues.redhat.com/browse/RFE-2236

More details on EFA can be found https://docs.aws.amazon.com/AWSEC2/latest/UserGuide/efa.html

According to the AWS API reference docs, we should just need to set the Interface Type on the EC2 Network Interface request to the value of `interface`, or `efa`. For now, I've left the PR keeping the two constants as matching the EC2 types, though if we think the names could be improved, I'm open to suggestions

### CODE REVIEW

File: machine/v1beta1/types_awsprovider.go
Comment (Fedosin) [inline machine/v1beta1/types_awsprovider.go:255]: Do we need 2 empty lines here?

File: machine/v1beta1/types_awsprovider.go
Comment (alexander-demichev) [inline machine/v1beta1/types_awsprovider.go:51]: I know we don't use openapi validations, but can we have an enum here, just to be consistent with what will be added in upstream CAPI?

File: machine/v1beta1/types_awsprovider.go
Comment (deads2k) [inline machine/v1beta1/types_awsprovider.go:50]: good sentence about omitted, can you add

```
	// Valid values are "interface", "efa", and omitted, which means no opinion and the platform
	// chooses a good default which may change over time.
	// The current default is <blah>
```

Can I also get a link to the AWS doc in here.  I know it may go stale, but let's at least try.

File: machine/v1beta1/types_awsprovider.go
Comment (deads2k) [inline machine/v1beta1/types_awsprovider.go:47]: do we make use of secondary network interfaces?  I don't mind only controlling this one, but I'd like to know.

Comment (JoelSpeed) [reply machine/v1beta1/types_awsprovider.go:47]: We only support a single network interface within Machine API across all of the core cloud providers, I'm not sure why this was done as this was pre me joining RH

**Comment by deads2k** [comment]
I don't see an enhancement linked here.  I think there are some questions to answer before we grown our API surface.

1. does the regularly consuming team know this is coming.  I think this is *mostly* the machine-api team.
2. does this work for masters.  My understanding is that masters are created by the installer, not the machine-api.  If this is the case, then the installer needs to review this API and determine how they will implement it.  We need an installer ACK before this can merge.
3. from the networking team, I'd like to understand if this new mode causes them an issues with debugging or constrains their future choices.
4. does the QE team know that this is adding to the list of configurations to test

These would all generally be answered in an enhancement.  if the installer says it doesn't impact them and the networking team says it doesn't impact them, then I'm ok to proceed without one if the machine-api team says its easy **and** promises to write an automated test for it that will block payload promotion.  We write enhancements to make this coordination easier.

While I can appreciate that some changes are more urgent than others, urgently delivering half a solution or a solution that does not formalize a means to ensure that it continues to work over time, our future selves will not be pleased.

/hold

oh, and the API itself looks pretty reasonable.

**Review by elmiko — COMMENTED** [review]
+1, generally looks good to me

**Comment by JoelSpeed** [comment]
Thanks for the feedback David, some notes in response to the questions

> does the regularly consuming team know this is coming. I think this is mostly the machine-api team.

This will primarily be a machine API feature, in which case myself and others from the team have already reviewed this

> does this work for masters. My understanding is that masters are created by the installer, not the machine-api. If this is the case, then the installer needs to review this API and determine how they will implement it. We need an installer ACK before this can merge.

I have tested the feature and there is no technical reason I have found for why this feature would break masters. You are correct that masters are created day 1 by the installer and as such, if we want to have support for this feature on masters, this will need to be added to the installer.

As far as I have been told, this feature request from the customer was for day 2 initially, as it is possible to reconfigure existing VMs to use the feature, but they want autoscaling via MAPI to avoid toil once the cluster is bootstrapped.

For the installer sake, I know they are very busy right now, but if they were to support this, the prerequisites are that the GO SDK and terraform both should support the feature.

Based on [the docs](https://registry.terraform.io/providers/hashicorp/aws/latest/docs/resources/network_interface#interface_type), the terrform provider has support and we should be able to include the option (also [the code](https://github.com/hashicorp/terraform-provider-aws/blob/d227cac8b6bc4a3cbd86acf16ac108d101b1d015/aws/data_source_aws_network_interface.go#L87) for the specific version)

Likewise, the [SDK Docs](https://pkg.go.dev/github.com/aws/aws-sdk-go@v1.35.20/service/ec2#InstanceNetworkInterfaceSpecification) for the version used by the installer also include the option.

I believe the only thing prohibiting the installer here would be a lack of time

> promises to write an automated test for it that will block payload promotion

Just to clarify, you would like to see a E2E blocking job that runs all machines with this option enabled to make sure all of our tests pass as normal?

**Comment by tkatarki** [comment]
right having masters also use efa instances and efa interfaces is not an ask for the mvp. @deads2k and @romfreiman

**Comment by romfreiman** [comment]
@danwinship anyone has experience with efa? Any thoughts on how sdn can be affected?

**Comment by danwinship** [comment]
> @danwinship anyone has experience with efa? Any thoughts on how sdn can be affected?

I don't know anything about it, but from the docs it seems like EFAs offer a superset of the functionality of ordinary network adapters, so _just_ switching the adapter type shouldn't break anything, hurt debuggability, or anything else.

The docs say "An EFA requires a security group that allows all inbound and outbound traffic to and from the security group itself." We don't currently have that, so something is going to have to make that happen when EFAs are enabled. AFAIK currently the security groups are created by the installer in an undocumented manner which is not guaranteed to remain unchanged in the future, so stuff will have to be figured out here. (And also of course, changing the security groups in this way means we'd be dropping all firewalling between worker nodes when using EFAs, meaning there will be less total network security, but maybe people who want to use EFAs don't care about that.)

It also seems like actually _using_ the extra EFA functionality might require extra kernel modules, sysctl changes, etc? I don't see anything in the RFEs about making sure that the HPC functionality is actually _usable_ with OCP when we enable EFAs, and if that hasn't happened, we should probably do that before adding APIs to enable them.

**Comment by JoelSpeed** [comment]
Just wanted to add a note, as I'm being poked, that we have pressure from PM (@tkatarki knows more) to try and get this feature merged ahead of FF even without demonstrating that it can be used to end to end. The suggestion was to merge this and the feature implementation under regression testing, rather than waiting for the full end to end flow to be demonstrated. I would assume that if we don't get it working by code freeze, we would revert the feature out.

> It also seems like actually using the extra EFA functionality might require extra kernel modules, sysctl changes, etc? I don't see anything in the RFEs about making sure that the HPC functionality is actually usable with OCP when we enable EFAs, and if that hasn't happened, we should probably do that before adding APIs to enable them.

There is work happening at the moment to get full end to end feature working with extra required software deployed via daemonsets, @kwozyman is giving regular updates on slack for these. He has also managed to verify that this API and associated machine patch work as expected and allowed him to continue his development of the full end to end feature

**Comment by JoelSpeed** [comment]
/hold cancel

The enhancement has been approved and merged, I think we can carry on with this and prepare it for merging post branch splits at the end of the week

**Review by alexander-demichev — APPROVED** [review]
/lgtm

**Comment by JoelSpeed** [comment]
Just to preempt a possible question about how this interacts with placement groups, IMO they should be considered separately. Each of the features can be used separately and there are no cross dependencies between the API types or PRs to implement each of the features.

The only reason these are linked at the moment is because we have a customer who is looking to get the absolute maximal performance from the EFA networking, for this, you can drop latency from approx 70-80us to 20us by using a cluster placement group with the EFA interface.

**Comment by JoelSpeed** [comment]
/label docs-approved
/label px-approved
/label qe-approved

We are not a no-FF team, so I'm adding these labels manually

**Pull request merged**
